runtime/service.go

```diff
@@ -1183,42 +1184,58 @@ func (s *service) shutdown(
 	shutdownCtx, cancel := context.WithTimeout(requestCtx, timeout)
 	defer cancel()
 	go func() {
-		// Once the shutdown procedure is done, the shim needs to shutdown too.
-		// This also ensures that if the VMM is still alive, it will receive a
-		// SIGKILL via exec.CommandContext
-		<-shutdownCtx.Done()
+		// Once the firecracker process is complete, the shim needs to shutdown too.
+		err := s.machine.Wait(context.Background())
+		if err != nil {
+			s.logger.WithError(err).Error("the VMM didn't finish successfully")
+		}
 		s.shimCancel()
 	}()
 
 	s.logger.Info("stopping the VM")
 
-	// Try to tell agent to exit, causing the VM to begin a reboot. If that
-	// fails, try to forcibly stop the VMM. If that too fails, just cancel
-	// the shutdownCtx to fast-path to the VMM getting SIGKILL.
-	_, shutdownErr := s.agentClient.Shutdown(shutdownCtx, req)
-	if shutdownErr != nil {
-		s.logger.WithError(shutdownErr).Error("failed to shutdown VM agent")
-		stopVMMErr := s.machine.StopVMM()
-		if stopVMMErr != nil {
-			s.logger.WithError(stopVMMErr).Error("failed to forcibly stop VMM")
-			cancel()
+	// 1) Tell agent to exit, causing the VM to begin a reboot.
+	_, err := s.agentClient.Shutdown(shutdownCtx, req)
+	if err != nil {
+		s.logger.WithError(err).Error("failed to shutdown VM agent")
+	} else {
+		// If the agent can take the request, wait a moment to gracefully shutdown.
+		err = s.machine.Wait(shutdownCtx)
+		if err != nil {
```

Comment:
TODO: I probably should handle context.Cancelled as well. Since we are shutting down the agent, the shim context (== firecracker process's context?) would be cancelled by the termination of the firecracker process.

Reply:
s.machine.Wait() is now taking context.Background() which won't be cancelled.